pallets/fusion/src/lib.rs

```diff
@@ -0,0 +1,1989 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+mod traits;
+mod types;
+mod weights;
+
+use crate::types::*;
+use frame_support::{
+	dispatch::GetDispatchInfo,
+	pallet_prelude::*,
+	traits::{
+		Currency, ExistenceRequirement, LockableCurrency, OnUnbalanced, UnfilteredDispatchable,
+		WithdrawReasons,
+	},
+	PalletId,
+};
+use frame_system::pallet_prelude::*;
+pub use pallet::*;
+use sp_core::U256;
+use sp_runtime::{
+	traits::{AccountIdConversion, Bounded, Zero},
+	Perbill, Saturating,
+};
+use sp_staking::EraIndex;
+pub use traits::{EraProvider, FusionExt};
+pub use weights::WeightInfo;
+
+pub type BalanceOf<T> =
+	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+pub type NegativeImbalanceOf<T> = <<T as Config>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::NegativeImbalance;
+
+pub const AVAIL_CURRENCY_ID: u32 = 0;
+pub const AVAIL_POOL_ID: u32 = 0;
+pub const MILLISECONDS_PER_YEAR: u64 = 1000 * 3600 * 24 * 36525 / 100;
+
+#[frame_support::pallet]
+pub mod pallet {
+
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// A sudo-able call.
+		type RuntimeCall: Parameter
+			+ UnfilteredDispatchable<RuntimeOrigin = Self::RuntimeOrigin>
+			+ GetDispatchInfo;
+
+		/// Type representing the weight of this pallet
+		type WeightInfo: WeightInfo;
+
+		/// Currency type for this pallet.
+		type Currency: LockableCurrency<Self::AccountId, Moment = BlockNumberFor<Self>>;
+
+		/// The destination for rewards that were not claimed after 'HistoryDepth' eras.
+		type RewardRemainder: OnUnbalanced<NegativeImbalanceOf<Self>>;
+
+		/// Pallet id used to derive accounts used by the pallet
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Maximum allowed for the currency name
+		#[pallet::constant]
+		type MaxCurrencyName: Get<u32>;
+
+		/// Maximum number of members in a pool
+		#[pallet::constant]
+		type MaxMembersPerPool: Get<u32>;
```

Comment:
Somewhat related: Do we foresee this needing to changeable by TC?

Reply:
Probably not, still changeable by runtime upgrade but we will see.
Pls resolve this if you have not more question